assert/equal.ts

```diff
@@ -6,71 +6,134 @@ function isKeyedCollection(x: unknown): x is KeyedCollection {
   return x instanceof Set || x instanceof Map;
 }
 
-function constructorsEqual(a: object, b: object) {
-  return a.constructor === b.constructor ||
-    a.constructor === Object && !b.constructor ||
-    !a.constructor && b.constructor === Object;
+function prototypesEqual(a: object, b: object) {
```

Comment:
Changed `constructorsEqual` to `prototypesEqual` because `obj.constructor` can easily be faked or inadvertently changed, whereas `Object.getPrototypeOf` can't (without monkey-patching `Object.getPrototypeOf` itself). This was necessitated due to some of the other changes affecting tests such as `assertFalse(equal(new WeakMap(), { constructor: WeakMap }));`